vhdl_syntax/src/parser/productions/names.rs

```diff
@@ -32,4 +74,480 @@ impl<T: TokenStream> Parser<T> {
+        self.separated_list(Parser::name, Comma);
+        self.end_node();
+    }
+
+    fn suffix(&mut self) {
+        // LRM §8.3
+        // suffix ::= identifier | string_literal | character_literal | `all` ;
+        self.expect_one_of_tokens([
+            Identifier,
+            StringLiteral,
+            CharacterLiteral,
+            Keyword(Kw::All),
+        ]);
+    }
+
+    fn name_tail(&mut self) {
+        // name             ::= prefix [ name_tail ] ;
+        // name_tail        ::= selected_name | attribute_name | indexed_name | slice_name | function_name ;
+        // selected_name    ::= `.` suffix [ name_tail ] ;
+        // attribute_name   ::= [ signature ] `'` identifier [ `(` expression `)` ] [ name_tail ] ;
+        // function_name    ::= `(` association_list `)` [ name_tail ] ;
+        // indexed_name     ::= `(` expression { `,` expression } `)` [ name_tail ] ;
+        // slice_name       ::= `(` discrete_range `)` [ name_tail ] ;
+
+        if self.next_is(Dot) {
+            self.start_node(SelectedName);
+            self.expect_token(Dot);
+            self.suffix();
+            self.end_node();
+            self.name_tail();
+        } else if self.next_is(LeftPar) {
+            // Try to differentiate between function calls, indexed names and slices as good as possible:
+            // 1. An `association_list` can be uniquely identified by searching for a '=>' inside the parenthesis
+            // 2. When at least a single comma is found, try parse the contents as an expression list
+            // 3. When the `to` or `downto` keyword is found, parse the content as a slice name
+            //
+            // If none of these apply, it can be either a `subtype_indication` or a single `expression`
+
+            if self.lookahead_in_parens([RightArrow]).is_some() {
+                self.start_node(FunctionCallOrIndexedName);
+                self.expect_token(LeftPar);
+                self.association_list();
+                self.expect_token(RightPar);
+                self.end_node();
+            } else if self.lookahead_in_parens([Comma]).is_some() {
+                self.start_node(FunctionCallOrIndexedName);
+                self.expect_token(LeftPar);
+                self.expression_list();
+                self.expect_token(RightPar);
+                self.end_node();
+            } else if self
+                .lookahead_in_parens([Keyword(Kw::To), Keyword(Kw::Downto)])
+                .is_some()
+            {
+                self.start_node(SliceName);
+                self.expect_token(LeftPar);
+                let closing_paren_distance = self.distance_to_closing_paren();
+                self.range(closing_paren_distance.unwrap());
+                self.expect_token(RightPar);
+                self.end_node();
+            } else {
+                // TODO: subtype_indication or expression?
+                self.start_node(Internal(SubtypeIndicationOrExpressionTokens));
```

Comment:
There may be a way to do the differentiation with limited lookahead, but I am a little afraid that it may become hard to reason about. That's something I can look into if you want. 

If you want to keep it simple, we could try to do the parsing of the parenthesized name tails in a second pass using backtracking?

Reply:
I think I prefer the simple solution. Maybe the second pass should also be pushed to the analysis stage when more information about the prefix is known. This would mimic the LRM.
My only preference is one can improve on this in the future without API breaking changes.

Reply:
Ok, should I get  rid of the differentiation of what's inside the parenthesis altogether and just generate a `RawTokens` node? Otherwise we will sometimes end up having an actual `SyntaxNode` while in other cases we have `RawTokens`. Not sure how useful it would be to a user.

Reply:
Yes, I think that makes sense at this point.